modules/system-disaster-recovery/src/main/java/org/apache/ignite/internal/disaster/system/MetastorageRepairImpl.java

```diff
@@ -0,0 +1,181 @@
+        this.logicalTopology = logicalTopology;
+        this.cmgManager = cmgManager;
+    }
+
+    @Override
+    public CompletableFuture<Void> repair(Set<String> participatingNodeNames, int metastorageReplicationFactor) {
+        LOG.info("Starting MG repair [participatingNodes={}, replicationFactor={}].", participatingNodeNames, metastorageReplicationFactor);
+
+        return waitTillValidatedNodesContain(participatingNodeNames)
+                .thenCompose(unused -> collectMetastorageIndexes(participatingNodeNames))
+                .thenCompose(indexes -> {
+                    LOG.info("Collected metastorage indexes [indexes={}].", indexes);
+
+                    Set<String> newMgNodes = nodesWithBestIndexes(indexes, metastorageReplicationFactor);
+                    LOG.info("Chose new MG nodes [mgNodes={}].", newMgNodes);
+
+                    String bestNodeName = chooseNodeWithBestIndex(indexes, newMgNodes);
+                    LOG.info("Chose best MG node [node={}].", bestNodeName);
+
+                    return cmgManager.changeMetastorageNodes(newMgNodes)
+                            .thenCompose(unused -> appointLeader(bestNodeName, indexes.get(bestNodeName).term(), newMgNodes))
+                            .thenRun(() -> LOG.info("Appointed MG leader forcefully [leader={}].", bestNodeName));
+                });
+    }
+
+    private CompletableFuture<Void> waitTillValidatedNodesContain(Set<String> nodeNames) {
+        Set<String> cumulativeValidatedNodeNames = ConcurrentHashMap.newKeySet();
+        CompletableFuture<Void> future = new CompletableFuture<>();
+
+        LogicalTopologyEventListener listener = new LogicalTopologyEventListener() {
+            @Override
+            public void onNodeValidated(LogicalNode validatedNode) {
+                cumulativeValidatedNodeNames.add(validatedNode.name());
+
+                if (contains(cumulativeValidatedNodeNames, nodeNames)) {
+                    future.complete(null);
+                }
+            }
+
+            @Override
+            public void onNodeInvalidated(LogicalNode invalidatedNode) {
+                cumulativeValidatedNodeNames.remove(invalidatedNode.name());
+            }
+        };
+
+        logicalTopology.addEventListener(listener);
+
+        cmgManager.validatedNodes()
+                .thenAccept(validatedNodes -> {
+                    Set<String> validatedNodeNames = validatedNodes.stream()
+                            .map(ClusterNode::name)
+                            .collect(toSet());
+                    if (contains(validatedNodeNames, nodeNames)) {
+                        future.complete(null);
+                    }
+
+                    cumulativeValidatedNodeNames.addAll(validatedNodeNames);
+                });
+
+        return future
+                .thenRun(() -> logicalTopology.removeEventListener(listener));
+    }
+
+    private static boolean contains(Set<String> container, Set<String> containee) {
```

Comment:
Let's rename to `containsAll`, `contains` usually means `∈`, not `⊂`

Reply:
I renamed it to `isSuperset()` as I would want to remain in the Maths terminology. Is it fine?

Reply:
Sure